contracts/test-utils/src/assertions.ts

```diff
@@ -159,10 +187,76 @@ export async function expectTransactionFailedWithoutReasonAsync(p: sendTransacti
  * @returns a new Promise which will reject if the conditions are not met and
  * otherwise resolve with no value.
  */
-export async function expectContractCallFailedAsync<T>(p: Promise<T>, reason: RevertReason): Promise<void> {
+export async function expectContractCallFailedAsync<T>(
+    p: Promise<T>,
+    reason: RevertReason | RichRevertReason,
+): Promise<void> {
+    if (reason instanceof RichRevertReason) {
+        return _expectContractCallRichRevertAsync<T>(p, reason);
+    }
     return expect(p).to.be.rejectedWith(reason);
 }
 
+/**
+ * A hacky way of expecting a rich revert reason to resolve.
+ * This is because geth does not signal when a revert happens during an
+ * eth_call so we're left to guess whether the returned data is an encoded
+ * rich revert reason.
+ * @param p A Promise resulting from a contract call.
+ * @param reason a rich revert reason
+ * @returns a promise which will reject `p` does not resolve to a rich revert reason.
+ */
+async function _expectContractCallRichRevertAsync<T>(p: Promise<T>, reason: RichRevertReason): Promise<void> {
+    let decoded: RichRevertReason | undefined;
+    try {
+        // Geth will not give any indication that a revert happened.
+        // It will return data as if it had succeeded, but
+        // the return data will be the abi encoded bytes of the revert data.
+        decoded = _tryToDecodeRichRevertReason(await p);
+    } catch (err) {
+        if (err.returnData) {
+            // We get here on geth if the contract wrapper fails to ABI decode the
+            // return data, which is a clue that it might be a rich revert reason.
+            decoded = _tryToDecodeRichRevertReason(err.returnData);
+        } else if (!err.results) {
+            // The contract wrapper caught a standard Error(string) revert reason
+            // so just wrap it in a StandardError.
+            decoded = new StandardError(err.message);
+        } else if (err.results) {
+            // Ganache threw a revert error.
+            const info = err.results[_.keys(err.results)[0]];
+            if (info.error === 'revert' && info.return) {
+                decoded = _tryToDecodeRichRevertReason(info.return);
+            }
+        }
+    }
+    if (decoded) {
```

Comment:
We try to be a bit tighter on our existence checks. This will not pass if `decoded` is set to `false`, `undefined` or `null`. Since we've decided to never use `null` in our codebase, we prefer to check for `undefined` explicitly:

```
if (decoded !== undefined) {
```

This catches unexpected values sooner rather then later.